### PR TITLE
Transaction tree tweaks

### PR DESCRIPTION
**Pull request by nhunzaker**
Second attempt...

This PR moves away from a graph like data structure and implements history as a tree. It's also incredibly fast:

```
Nodes  ::append()  ::root()  ::reduce(merge)  ::size()  ::prune()  Memory Growth
-----  ----------  --------  ---------------  --------  ---------  -------------
10000  0.0008ms    0.40ms    1.81ms           0.32ms    1.24ms     0.883%

Dispatching 10000 actions: 14ms
```

Performance is Similar in Safari and slightly less stellar in Firefox. In any case: 10,000 unique transactions processed is a very lofty goal. This is essentially pushing a new action ever 16 milliseconds for 2.6 minutes. If we ever want to enable high fidelity rewind, we can definitely do it...

undo-tree

### CODE REVIEW

File: examples/undo-tree/index.jsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import Microcosm from 'Microcosm'
+import React from 'react'
+import Drawing from './components/drawing'
+
+class UndoTree extends Microcosm {
+
+  shouldHistoryKeep(transaction) {
+    return false
+  }
```
Comment (nhunzaker) [inline examples/undo-tree/index.jsx]: Thoughts on this name, @efatsi?

File: examples/undo-tree/index.jsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import Microcosm from 'Microcosm'
+import React from 'react'
+import Drawing from './components/drawing'
+
+class UndoTree extends Microcosm {
+
+  shouldHistoryKeep(transaction) {
+    return true
+  }
```
Comment (nhunzaker) [inline examples/undo-tree/index.jsx]: Sorry to repost... bad push. @efatsi thoughts?

Comment (efatsi) [inline examples/undo-tree/index.jsx]: sounds like a good name to me. whatever the name though, this logic is not trivial and deserves a good writeup in the readme i believe.

Comment (nhunzaker) [inline examples/undo-tree/index.jsx]: I agree.

Comment (nhunzaker) [inline examples/undo-tree/index.jsx]: Really, I continue to struggle for a name here. It's "should the app commit this transaction into base state and purge it from history"

Comment (efatsi) [inline examples/undo-tree/index.jsx]: you could always go the verbose route `shouldMergeTransactionIntoBaseStateAndPurgeFromHistory`

Comment (nhunzaker) [inline examples/undo-tree/index.jsx]: Haha, you actually beat me to typing that exact thing out!

**Pull request merged**